scripts/check_test_inclusion.py

```diff
@@ -0,0 +1,89 @@
+#!/bin/python
+
+"""
+Verifies that all tests are a part of the project file.
+"""
+
+from __future__ import print_function
+import os
+import os.path
+import re
+import sys
+
+
+# Tests that are known not to compile in Xcode and can't be added there.
+EXCLUDED = frozenset([
+    # b/79496027
+    'Firestore/core/test/firebase/firestore/remote/serializer_test.cc',
+])
+
+
+def Main(args):
+    problems = CheckProject('Firestore/Example/Firestore.xcodeproj/project.pbxproj',
+                            'Firestore/Example/Tests', 'Firestore/core/test')
+
+    problems = FilterProblems(problems)
+    problems.sort()
+
+    if len(problems) > 0:
+        Error('Test files exist that are unreferenced in Xcode project files:')
+        for problem in problems:
+            Error(problem)
+        sys.exit(1)
+
+    sys.exit(0)
+
+
+def CheckProject(project_file, *test_dirs):
+    test_files = FindTestFiles(*test_dirs)
+    basenames = MakeBasenames(test_files)
+
+    file_list_pattern = re.compile(r'/\* (\S+) in Sources \*/')
+    with open(project_file, 'r') as fd:
+        for line in fd:
+            line = line.rstrip()
+            m = file_list_pattern.search(line)
+            if m:
+                basename = m.group(1)
+                if basename in basenames:
+                    del basenames[basename]
+
+    return sorted(basenames.values())
+
+
+def FindTestFiles(*srcroots):
+    result = []
+    for srcroot in srcroots:
+        for root, dirs, files in os.walk(srcroot):
+            for file in files:
+                result.append(os.path.join(root, file))
+    return result
+
+
+def MakeBasenames(filenames):
+    test_file_pattern = re.compile(r'(?:Tests?\.mm?|_test\.(?:cc|mm))$')
+    result = dict()
+    for filename in filenames:
+        basename = os.path.basename(filename)
+        m = test_file_pattern.search(basename)
+        if m:
+            result[basename] = filename
+
+    return result
+
+
+def FilterProblems(problems):
+    result = []
+    for problem in problems:
+        if problem not in EXCLUDED:
+            result.append(problem)
+    return result
```

Comment:
More "pythonic" alternatives:

old: Use filter(), i.e.:
```
return filter(lambda problem : problem not in EXCLUDED, problems)
```

Use list comprehension:
```
return [problem for problem in problems if problem not in EXCLUDED];
```

Use generator (though since you sort the result, this looses a bit of it's appeal):
```
def FilterProblems(problems):
  for problem in problems:
    if problem not in EXCLUDED: yield problem
```

Optional.

Note: This also applies to some of the loops above (in the rest of the file.) If you make a change here, consider applying it throughout (though obviously apply judgement first as things like list comprehension become rapidly unreadable if taken too far.)

Reply:
I added this check directly to `FindTestFiles`. There's no need to process a source file we're going to exclude anyway.